src/TreeSitter/Importing.hs

```diff
@@ -1,135 +1,271 @@
+peekFieldName :: (Carrier sig m, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (Maybe FieldName)
+peekFieldName = do
+  cursor <- ask
+  fieldName <- liftIO $ ts_tree_cursor_current_field_name cursor
+  if fieldName == nullPtr then
+    pure Nothing
+  else
+    Just . FieldName <$> liftIO (peekCString fieldName)
+
+-- | Return the fields remaining in the current branch, represented as 'Map.Map' of 'FieldName's to their corresponding 'Node's.
+getFields :: (Carrier sig m, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (Map.Map FieldName Node)
+getFields = go Map.empty >>= \fields -> liftIO (print (Map.keys fields)) >> pure fields
+  where go fs = do
+          node <- peekNode
+          case node of
+            Just node' -> do
+              fieldName <- peekFieldName
+              keepGoing <- step
+              let fs' = case fieldName of
+                    Just fieldName' -> Map.insert fieldName' node' fs
+                    _ -> fs
+              if keepGoing then go fs'
+              else pure fs'
+            _ -> pure fs
 
--- | Return a 'ByteString' that contains a slice of the given 'Source'.
+-- | Return a 'ByteString' that contains a slice of the given 'ByteString'.
 slice :: Int -> Int -> ByteString -> ByteString
 slice start end = take . drop
   where drop = B.drop start
         take = B.take (end - start)
 
-class Importing type' where
 
-  import' :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m type'
+newtype FieldName = FieldName { getFieldName :: String }
+  deriving (Eq, Ord, Show)
+
+
+-- | Generic construction of ASTs from a 'Map.Map' of named fields.
+--
+--   Product types (specifically, record types) are constructed by looking up the node for each corresponding field name in the map, moving the cursor to it, and then invoking 'buildNode' to construct the value for that field. Leaf types are constructed as a special case of product types.
+--
+--   Sum types are constructed by attempting to build each constructor nondeterministically. This should instead use the current node’s symbol to select the corresponding constructor deterministically.
+class GBuilding f where
+  gbuildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (f a)
+-- we'd only build the map when we know we're looking at a product
+
+
+instance GBuilding f => GBuilding (M1 D c f) where
+  gbuildNode = M1 <$> gbuildNode -- current node, not first child like above in original Building definition
+  -- gSymbolMatch _ = gSymbolMatch (Proxy @f)
+
+instance GBuilding f => GBuilding (M1 C c f) where
+  gbuildNode = M1 <$> gbuildNode
+
+-- Possibly for anonymous leaf nodes:
+instance GBuilding U1 where
+  gbuildNode = pure U1
+
+-- For regular leaf nodes
+instance {-# Overlappable #-} GBuilding (M1 S s (K1 c Text.Text)) where
```

Comment:
Why do we need the Overlappable pragmas here? (Also, small stylistic nit: pragmas should be capitalized, e.g. `OVERLAPS` and `OVERLAPPABLE`.

Reply:
It’s cos this instance overlaps unavoidably with the one for unary products.

Reply:
Because we want the compiler to use `Text.Text` here in the case of regular leaf nodes, but if you look at the unary product `instance {-# Overlappable #-} (Selector s, Building k) => GBuilding (M1 S s (K1 c k))`, we have some polymorphic `k` that we _never_ want as `Text.Text`.